### PR TITLE
Report more detail in VM start errors

### LINKED ISSUE #397: VM launch failure doesn't report a useful error
If the VM fails to launch, the error reported to the user is a timeout after 10,000ms rather than an exit status or anything from the stdout or stderr. This makes it very hard to debug launch failures when using the integrated console. If using a terminal, the errors are visible in the terminal.

`AdvancedLaunchingConnector` doesn't notice that the VM process has terminated and instead has to wait to catch the `TransportTimeoutException` and then it reports a timeout.

I would love to improve this!

After the timeout we could check if the process has terminated (`process.isAlive()`) and report that exit status instead of a timeout, and read from the stdout and stderr to construct a better error message and log the stdout / stderr to the console so the user can see what happened. The `VMStartException` doesn't seem very well suited to our purpose as although it includes the `Process`, because we `.destroy()` it we don't have access to the streams. Perhaps we can make our own exception here that can glean more information from the `Process`.

Ideally instead of waiting for timeout we could observe the process termination early, so the user isn't left hanging for 10s. But it doesn't seem like that's easily supported in the underlying Eclipse code.

I'm going to tinker with this, but I'd love to know if this sounds like the right track for improvement.

### PR DESCRIPTION
**Pull request by karlvr**
Resolves #397 or at least contributes to it

### CODE REVIEW

**Comment by testforstephen** [comment]
This is a nice improvement.

Currently the CI reports some checkstyle errors, you can run checkstyle check via cmd `./mvnw checkstyle:check` in local machine.
```
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[1] (header) RegexpHeader: Line does not match expected header line of '^/\*+$'.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[10] (regexp) RegexpSingleline: Line has trailing spaces.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[10,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[11] (indentation) Indentation: 'member def type' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[11,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[12] (indentation) Indentation: 'member def type' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[12,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[13] (indentation) Indentation: 'member def type' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[13,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[14] (indentation) Indentation: 'member def type' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[14,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[16] (indentation) Indentation: 'ctor def modifier' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[16,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[17] (indentation) Indentation: 'ctor def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[17,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[18] (indentation) Indentation: 'ctor def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[18,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[19] (indentation) Indentation: 'ctor def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[19,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[20] (indentation) Indentation: 'ctor def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[20,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[21] (indentation) Indentation: 'ctor def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[21,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[22] (indentation) Indentation: 'ctor def rcurly' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[22,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[24] (indentation) Indentation: 'method def modifier' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[24,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[25] (indentation) Indentation: 'method def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[25,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[26] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[26,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[28] (indentation) Indentation: 'method def modifier' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[28,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[29] (indentation) Indentation: 'method def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[29,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[30] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[30,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[32] (indentation) Indentation: 'method def modifier' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[32,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[33] (indentation) Indentation: 'method def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[33,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[34] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[34,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[36] (indentation) Indentation: 'method def modifier' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[36,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[37] (indentation) Indentation: 'method def' child has incorrect indentation level 16, expected level should be 8.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[37,1] (whitespace) FileTabCharacter: Line contains a tab character.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[38] (indentation) Indentation: 'method def rcurly' has incorrect indentation level 8, expected level should be 4.
Error:  src/main/java/com/microsoft/java/debug/core/LaunchException.java:[38,1] (whitespace) FileTabCharacter: Line contains a tab character.

```

**Comment by karlvr** [comment]
@testforstephen thank you. Just a few checkstyle errors there! I've force-pushed a fix for them to keep the commit history clean.

I have been investigating terminating early if the process exits... I don't know if there are any clever tricks to do that here, I haven't done any work with `Future`s! I'm sort of looking for a "race" between the `Process` exiting and the debug connection `accept` returning... also given that `accept` will timeout. Any suggestions welcome otherwise I'll continue to experiment.

**Comment by testforstephen** [comment]
https://github.com/microsoft/java-debug/blob/e6655ead412ceed5afa37b3781ed84e0b8ba425a/com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java#L86-L95

How about using `process.onExit()` to watch the exit event and return early?
```java
        process.onExit().whenCompleteAsync((p, ex) -> {
            if (vm == null) {  // When the debug connection is not set up yet.
                  listenConnector.stopListening(args);
                  ...
            }
        });
```

**Comment by karlvr** [comment]
@testforstephen I have pushed some additional work that detects early termination of the launched process, so we no longer have to wait for the timeout to report the error.

**Comment by testforstephen** [comment]
Just came back from vacation, I will find some time this week to review this PR?

BTW, do you have any sample project to help reproduce the issue at https://github.com/microsoft/java-debug/issues/397? thanks.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java [hunk 1]
```diff
@@ -83,18 +90,100 @@ public VirtualMachine launch(Map<String, ? extends Argument> connectionArgs)
...
+        /* Launch the Java process */
+        final Process process = Runtime.getRuntime().exec(cmds, envVars, workingDir);
+
+        /* A Future that will be completed if we successfully connect to the launched process, or
+           will fail with an Exception if we do not.
+         */
+        final CompletableFuture<VirtualMachineImpl> result = new CompletableFuture<>();
+
+        /* Listen for the debug connection from the Java process */
+        ForkJoinPool.commonPool().execute(() -> {
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java:103]: Prefer to use CompletableFuture.runAsync(...)

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java [hunk 2]
```diff
@@ -83,18 +90,100 @@ public VirtualMachine launch(Map<String, ? extends Argument> connectionArgs)
...
+                    stdout,
+                    stderr
+                ));
+            } catch (RuntimeException e) {
+                result.completeExceptionally(e);
+            }
+        });
+
+        /* Wait for the Java process to exit; if it exits before the debug connection is made, report it as an error. */
+        process.onExit().thenAcceptAsync(theProcess -> {
```
[the remark this answers is not in the document]
Comment (karlvr) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java:134]: @testforstephen I'm not sure. I have seen socket-related things not detecting disconnection etc. I tried to tidy everything up in each case. Is there a particular case that you're thinking may be overkill?

**Comment by karlvr** [comment]
@testforstephen I hope you had a great vacation. I have added a very simple example of reproducing the error to #397.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java:134]: I see process.onExit() is hit immediately with your sample. Looks good.

**Review by testforstephen — APPROVED** [review]
LGTM. @karlvr thanks for the contribution.

**Pull request merged**